MIP-43: AB-FFS Governed Gas Pool (#44)
movementlabsxyz/MIP: MIP/mip-n/README.md

> # MIP-42: AB-FFS Governed Gas Pool

Reviewer: ```suggestion
# MIP-42: Governed Gas Pool
```
This is no longer coupled with AB or AB-FFS

> # MIP-42: AB-FFS Governed Gas Pool

Reviewer 2: extends beyond swap-bridge-swap FFS

- **Description**: Introduces a governed gas pool for AB-FFS which allows for initial period wherein used gas is not recirculated.

> - **Authors**: [Liam Monninger](mailto:liam@movementlabs.xyz)

Reviewer: ```suggestion
- **Authors**: [Liam Monninger, Richard Melkonian](mailto:liam@movementlabs.xyz, mailto:richard@movementlabs.xyz)
```

- **Reviewer**: Andreas Penzkofer
- **Desiderata**: [MD-38](https://github.com/movementlabsxyz/MIP/pull/38)

## Abstract
We propose a governed gas pool both as provisional and potentially permanent solution to the supply problem described in [MD-38](https://github.com/movementlabsxyz/MIP/pull/38). This solution consists of depositing funds into a pool on the L2, and holding votes against actions for the recirculation of said funds.

> The concept of a governed gas pool can be applied generally to manage migrations between systems. For example, the governed gas pool could be used to transition from the "Trickle-back" mechanism referenced in [MD-38](https://github.com/movementlabsxyz/MIP/pull/38) to [MIP-42](https://github.com/movementlabsxyz/MIP/pull/42), as is used as an example amongst others herein.

Reviewer: ```suggestion
The concept of a governed gas pool can be applied generally to manage migrations between systems. For example, the governed gas pool could be used to transition from the "Trickle-back" mechanism referenced in [MD-38](https://github.com/movementlabsxyz/MIP/pull/38) to [MIP-42](https://github.com/movementlabsxyz/MIP/pull/42), as is used as an example amongst others herein. As demonstrated in the diagram, the governed gas pool can be used to provide liquidity for a variety of protocol operations, liquidity of L1 Rewards, Trickle Back liquidity or grants. To use the funds for any of these activities a proposal would have to pass and a network upgrade or migration would also have to occur after governance approves the proposal. It is recommended to read generally around [aptos-governance](https://aptos.web3doc.top/concepts/governance/) and how it is implemented. Because of its tie to aptos-staking (which we do not use), we will be required to make modifications to `aptos_governance.move`. Changes for this are suggested and laid in MIP(##).
```

## Motivation

> The AB-FFS under fixed supply presents several challenges presented in [MD-38](https://github.com/movementlabsxyz/MIP/pull/38). This MIP seeks to address the question of where used gas funds should be deposited with a simple and flexible solution. By simply depositing funds into a pool and holding votes against actions for the recirculation of said funds, we can create a system that can provide a short-term solution to the supply problem and generalize to other systems. 

Reviewer 2: ```suggestion
The requirement for fixed supply presents several challenges and where the challenges dependent on the selected design, see Appendix A1 in [MD-38](https://github.com/movementlabsxyz/MIP/pull/38). One particular concern is that if fees are burnt they would decrease the supply. Hence fees, should be collected instead. 

This MIP seeks to address the question of where used gas funds should be deposited with a simple and flexible solution. By simply depositing funds into a pool and holding votes against actions for the recirculation of said funds, we can create a system that can provide a short-term solution to the supply problem and generalize to other systems. 
```

> The AB-FFS under fixed supply presents several challenges presented in [MD-38](https://github.com/movementlabsxyz/MIP/pull/38). This MIP seeks to address the question of where used gas funds should be deposited with a simple and flexible solution. By simply depositing funds into a pool and holding votes against actions for the recirculation of said funds, we can create a system that can provide a short-term solution to the supply problem and generalize to other systems. 

Reviewer: ```suggestion
The AB-FFS under fixed supply presents several challenges presented in [MD-38](https://github.com/movementlabsxyz/MIP/pull/38). This MIP seeks to address the question of where used gas funds should be deposited with a simple and flexible solution. By simply depositing funds into a pool and holding votes against actions for the recirculation of said funds, we can create a system that can provide a solution to the supply problem and generalize to other systems. 
```

## Specification
![Governed Gas Pool](governed-gas-pool.png)

> 1. Gas used on the L2 MUST be deposited into the Governed Gas Pool.

Reviewer 2: Governed Gas Pool is a new term. its definition should be added to the Glossary and this MD should be referenced from Glossary, see 
https://github.com/movementlabsxyz/MIP/pull/45

2. The Governed Gas Pool MUST disperse deposited funds according to a proposal made by Proposers which is enacted by a vote of Executors.

> 3. The Governed Gas Pool SHOULD define a maximum emissions rate for the pool s.t. the governing body cannot engage in Byzantine or other malicious behavior.

Reviewer: ```suggestion
3. The Governed Gas Pool SHOULD define a maximum emissions rate for the pool s.t. the governing body cannot engage in Byzantine or other malicious behavior.
4. Governance MUST be implemented using the [aptos_governance](https://github.com/movementlabsxyz/aptos-core/blob/andyjsbell/atomic-bridge/aptos-move/framework/aptos-framework/sources/aptos_governance.move) module provided by aptos core. 
5. Any necessary upgrades or migrations required to `aptos_framework` MUST be included within a proposal and voted upon by the governing body. 
```

Author: Governed gas pool is not specific to Aptos that would be a separate MIP.

Reviewer: Ok, in that case I'll put the reference implementation there also, as the ref implementation I wrote assumes usage of `aptos_governance.move`.

### Proposers and Executors
The Governed Gas Pool will have two roles: Proposers and Executors. Proposers will be able to propose actions for the pool, and Executors will be able to vote on these actions. 

The rationale behind these two separate roles is to encapsulate the decision-making process. This separation of concerns can allow off-chain services which are less-secure to propose actions which do not immediately affect the pool. Executors, who are more secure, can then vote on these actions.

### Actions
1. For simplicity, we suggest that all synchronous and atomic actions performed by the pool SHOULD be formulated as dispersals from the pool into other accounts. 

2. For security, we propose the constraint that all of these dispersals SHOULD be $O(1)$. 

> For example, the pool can securely--without risk of gas attack--disperse into [MIP-42](https://github.com/movementlabsxyz/MIP/pull/42) withholdings if said action only involves depositing funds into the account controlling that account. In the case that this action were linear in the size of the approved liquidity providers, for example, gas pool proposals good be subject to gas attacks which would prevent critical actions for recirculating funds to take place at the appropriate time.

Reviewer 2: what is liquidity providers job? liquidity providers are not yet defined in this document

Author: This is only alluded to in the diagram, but the governed gas pool could disperse to liquidity providers elsewhere in the network.

Reviewer 2: thx. but the role of the liquidity providers is still not defined nor clear.

> 3. When considering mechanisms which are not synchronous and atomic, we propose that the Governed Gas Pool SHOULD only be considered responsible for the security of the synchronous and atomic step constituting the initial dispersal. 

Reviewer 2: this is not clear. please elaborate in the above line

Author: Does the example below not clarify?

Reviewer 2: so why "synchronous". what does synchronous mean in this context. can synchronous be removed?

> For example, the "Trickle-back" mechanism referenced in [MD-38](https://github.com/movementlabsxyz/MIP/pull/38) could be performed with an initial dispersal into the bridge pool. However, the ultimate effect of those funds returning to L1 settlement rewards would not be guaranteed once set dispersal occurs. Thus, when analyzing the security of the "Trickle-back" mechanism, one would consider the actions of the Governed Gas Pool as secure insofar as the dispersal into the bridge pool is secure. The remainder of the mechanism would be considered separately.

Reviewer 2: ```suggestion
For example, the "Trickle-back" mechanism referenced in [MD-38](https://github.com/movementlabsxyz/MIP/pull/38) could be performed with an initial dispersal through the bridge (or into the L2 primary bridge DEX incase of the swap-bridge-swap solution). However, the ultimate effect of those funds returning to L1 settlement rewards would not be guaranteed once set dispersal occurs. Thus, when analyzing the security of the "Trickle-back" mechanism, one would consider the actions of the Governed Gas Pool as secure insofar as the dispersal into the bridge pool is secure. The remainder of the mechanism would be considered separately.
```

### Recommendations for Governance of the Governed Gas Pool

> We provide a series of recommendations for the governance of the Governed Gas Pool to account for common phenomena in this system. More detailed proposals, however, such be set forth in MIPs governing a specific Governed Gas Pool implementation.

Reviewer 2: missing?

Author: yes

### Migrating via the Governed Gas Pool
The Governed Gas Pool can be used to manage migrations between systems. For example, the Governed Gas Pool could be used to transition from the "Trickle-back" mechanism referenced in [MD-38](https://github.com/movementlabsxyz/MIP/pull/38) and [MIP-42](https://github.com/movementlabsxyz/MIP/pull/42)

## Reference Implementation

> 

Reviewer: This example module adheres to governance rules by allowing only aptos_governance to add proposals and only an authorized admin to execute them.

```rust
//pseudocode
module GovernedGasPool::goverened_gas_pool {
    use aptos_framework::aptos_coin::AptosCoin;
    use aptos_framework::coin;
    use aptos_framework::signer;
    use aptos_framework::aptos_governance;
    use aptos_std::vector;
    use aptos_std::option::{self, Option};

    // Address of the governance framework module
    const GOVERNANCE_ADDRESS: address = @0x1;

    struct GovernedPool has key {
        funds: coin::Coin<AptosCoin>,                // holds the pool of funds in AptosCoin
        passed_proposals: vector::Vector<Proposal>,  // list of proposals that passed governance
        admin: address                               // address with permission to execute proposals
    }

    struct DispersalAction has copy, drop, store {
        recipient: address,
        amount: u64
    }

    struct Proposal has key {
        id: u64,
        dispersal_action: Option<DispersalAction>, // Only one type of action at a time
        executed: bool,
    }

    /// Initialize the Governed Pool
    public fun initialize_governed_pool(admin: &signer, authorized_admin: address): address {
        let governed_pool = GovernedPool {
            funds: coin::zero<AptosCoin>(),
            passed_proposals: vector::empty<Proposal>(),
            admin: authorized_admin
        };
        let addr = signer::address_of(admin);
        move_to(admin, governed_pool);
        addr
    }

    /// Add a proposal that has passed governance
    /// Only callable by the governance module at address 0x1
    public fun add_passed_proposal(
        pool: &mut GovernedPool,
        governance_signer: &signer,
        id: u64,
        dispersal_action: DispersalAction
    ) {
        // Ensure only the governance framework can call this function
        assert!(signer::address_of(governance_signer) == GOVERNANCE_ADDRESS, 1);

        let proposal = Proposal {
            id,
            dispersal_action: option::some(dispersal_action),
            executed: false,
        };
        vector::push_back(&mut pool.passed_proposals, proposal);
    }

    /// Execute a proposal that has been approved by governance
    /// Only callable by the designated admin
    public fun execute_passed_proposal(pool: &mut GovernedPool, executor: &signer, proposal_id: u64) {
        // Ensure only the authorized admin can call this function
        assert!(signer::address_of(executor) == pool.admin, 2);

        let proposal_index = find_proposal(&pool.passed_proposals, proposal_id);
        let proposal = &mut vector::borrow_mut(&mut pool.passed_proposals, proposal_index);

        // Ensure the proposal hasn't already been executed
        assert!(!proposal.executed, 3);

        // Execute the dispersal action if present
        if (option::is_some(&proposal.dispersal_action)) {
            let action = option::borrow(&proposal.dispersal_action).unwrap();
            let amount = action.amount;
            assert!(amount <= coin::value(&pool.funds), 4); // Ensure pool has sufficient funds
            coin::withdraw(&mut pool.funds, amount);
            coin::deposit(&signer::create(action.recipient), amount);
        };

        proposal.executed = true;
    }
}
```

Reviewer: Moving this into MIP#48
